lib/srv/regular/sshserver.go

```diff
@@ -2117,6 +2324,117 @@ func (s *Server) handleProxyJump(ctx context.Context, ccx *sshutils.ConnectionCo
+func (s *Server) createForwardingContext(ctx context.Context, ccx *sshutils.ConnectionContext, r *ssh.Request) (context.Context, *srv.ServerContext, error) {
+	req, err := sshutils.ParseTCPIPForwardReq(r.Payload)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	identityContext, err := s.authHandlers.CreateIdentityContext(ccx.ServerConn)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	// On regular server in "normal" mode "tcpip-forward" requests from
+	// SessionJoinPrincipal should be rejected, otherwise it's possible to use
+	// the "-teleport-internal-join" user to bypass RBAC.
+	if identityContext.Login == teleport.SSHSessionJoinPrincipal {
+		log.Errorf("Request rejected, %q with SessionJoinPrincipal in forward node must be blocked", r.Type)
+		err := trace.AccessDenied("attempted %q request in join-only mode", r.Type)
+		if replyErr := r.Reply(false, []byte(utils.FormatErrorWithNewline(err))); replyErr != nil {
+			s.Logger.Warnf("Failed to reply to %q request: %v", r.Type, err)
+		}
+		// Disable default reply by caller, we already handled it.
+		r.WantReply = false
+		return nil, nil, err
+	}
+
+	// Create context for this request.
+	ctx, scx, err := srv.NewServerContext(ctx, ccx, s, identityContext)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	listenAddr := sshutils.JoinHostPort(req.Addr, req.Port)
+	scx.IsTestStub = s.isTestStub
+	scx.ExecType = teleport.TCPIPForwardRequest
+	scx.SrcAddr = listenAddr
+	scx.DstAddr = ccx.NetConn.RemoteAddr().String()
+	scx.SetAllowFileCopying(s.allowFileCopying)
+
+	if err := s.canPortForward(scx); err != nil {
+		scx.Close()
+		return nil, nil, trace.Wrap(err)
+	}
+	return ctx, scx, nil
+}
+
+// handleTCPIPForwardRequest handles remote port forwarding requests.
+func (s *Server) handleTCPIPForwardRequest(ctx context.Context, ccx *sshutils.ConnectionContext, r *ssh.Request) error {
+	ctx, scx, err := s.createForwardingContext(ctx, ccx, r)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer scx.Close()
+	listener, err := s.listenTCPIP(scx, scx.SrcAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Set the src addr again since it may have been updated with a new port.
+	scx.SrcAddr = listener.Addr().String()
+	event := scx.GetPortForwardEvent()
+	if err := s.EmitAuditEvent(ctx, &event); err != nil {
+		s.Logger.WithError(err).Warn("Failed to emit audit event.")
+	}
+	if err := sshutils.StartRemoteListener(ctx, scx.ConnectionContext.ServerConn, scx.SrcAddr, scx.DstAddr, listener); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Report addr back to the client.
+	if r.WantReply {
+		var payload []byte
+		req, err := sshutils.ParseTCPIPForwardReq(r.Payload)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if req.Port == 0 {
+			payload = make([]byte, 4)
+			_, port, err := sshutils.SplitHostPort(listener.Addr().String())
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			binary.BigEndian.PutUint32(payload, port)
```

Comment:
Consider changing this to a `ssh.Marshal` of a `struct { DeterminedPort uint32 }` or something.